comp/languagedetection/client/client.go

```diff
@@ -0,0 +1,328 @@
+			c.telemetry.Running.Set(1)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+// processEvent processes events from workloadmeta
+func (c *client) processEvent(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.logger.Tracef("Processing %d events", len(evBundle.Events))
+	for _, event := range evBundle.Events {
+		switch event.Entity.GetID().Kind {
+		case workloadmeta.KindProcess:
+			c.processProcessEvent(event, false)
+		case workloadmeta.KindKubernetesPod:
+			c.processPodEvent(event)
+		}
+
+	}
+}
+
+// startStreaming retrieves the language detection client (= the DCA Client) and periodically sends data to the Cluster-Agent
+func (c *client) startStreaming() {
+	freshUpdateTimer := time.NewTicker(c.newUpdatePeriod)
+	defer freshUpdateTimer.Stop()
+
+	periodicFlushTimer := time.NewTicker(c.periodicalFlushPeriod)
+	defer periodicFlushTimer.Stop()
+
+	if c.langDetectionCl == nil {
+		// TODO(components): The ClusterAgentClient should most likely be a component. Moreover it should provide a retry mechanism or at least, the duration before the next try.
+		// Since currently we never retry `GetClusterAgentClient` in other parts of the code, we choose to follow the same pattern.
+		dcaClient, err := clusteragent.GetClusterAgentClient()
+		if err != nil {
+			c.logger.Errorf("failed to get dca client %s, stopping language exporter", err)
+			c.cancel()
+			return
+		}
+		c.langDetectionCl = dcaClient
+	}
+
+	for {
+		select {
+		// frequently send only fresh updates
+		case <-freshUpdateTimer.C:
+			data := c.getFreshBatchProto()
+			c.send(data)
+		// less frequently, send the entire batch
+		case <-periodicFlushTimer.C:
+			data := c.getCurrentBatchProto()
+			c.send(data)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+// send sends the data to the cluster-agent. It doesn't implement a retry mechanism because if the dca is available
+// then the data will eventually be transmitted by the periodic flush mechanism.
+func (c *client) send(data *process.ParentLanguageAnnotationRequest) {
+	if data == nil {
+		return
+	}
+	t := time.Now()
+	err := c.langDetectionCl.PostLanguageMetadata(c.ctx, data)
+	if err != nil {
+		c.logger.Errorf("failed to post language metadata %v", err)
+		c.telemetry.Requests.Inc(statusError)
+		return
+	}
+	c.telemetry.Latency.Observe(time.Since(t).Seconds())
+	c.telemetry.Requests.Inc(statusSuccess)
+}
+
+// retryProcessEventsWithoutPod processes a second time process events for which the associated
+// pod was not found because it is possible that the pod will be added to workloadmeta after the
+// kubelet collector pulls data
+func (c *client) retryProcessEventsWithoutPod() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	for _, event := range c.processesWithoutPod {
+		c.processProcessEvent(event, true)
+	}
+	c.processesWithoutPod = make([]workloadmeta.Event, 0)
+}
+
+// processProcessEvents updates the current batch and the freshlyUpdatedPods
```

Comment:
To avoid the repeated pattern, you can call it `handleProcessEvents` and `handlePodEvent`